crates/sparrow-merge/src/merge_pipeline.rs

```diff
@@ -0,0 +1,356 @@
+use std::collections::VecDeque;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+
+use arrow_schema::DataType;
+use error_stack::{IntoReport, IntoReportCompat, ResultExt};
+use itertools::Itertools;
+use parking_lot::Mutex;
+use sparrow_batch::Batch;
+use sparrow_physical::StepId;
+use sparrow_scheduler::{
+    InputHandles, Partition, Partitioned, Pipeline, PipelineError, Scheduler, TaskRef,
+};
+
+use crate::gather::Gatherer;
+use crate::merge::BinaryMergeInput;
+use crate::spread::Spread;
+
+/// Runs a merge operation.
+pub struct MergePipeline {
+    /// The state for each partition.
+    partitions: Partitioned<Mutex<MergePartition>>,
+    /// Input handles for the consuming (receiving) computations.
+    consumers: InputHandles,
+    /// The id for the left input
+    input_l: StepId,
+    /// The id for the right input
+    input_r: StepId,
+    /// The left data type
+    datatype_l: DataType,
+    /// The right data type
+    datatype_r: DataType,
+    /// Keeps track of closed inputs.
+    is_closed_l: Mutex<bool>,
+    is_closed_r: Mutex<bool>,
+    /// The result type of this merge
+    result_type: DataType,
+}
+
+impl std::fmt::Debug for MergePipeline {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("MergePipeline").finish()
+    }
+}
+
+struct MergePartition {
+    /// Whether this partition is closed.
+    is_closed: AtomicBool,
+    /// Task for this partition.
+    task: TaskRef,
+    /// Inputs (for either side) for this partition.
+    ///
+    /// the `usize` should be either 0 (left), or 1 (right).
+    inputs: Mutex<VecDeque<(Batch, usize)>>,
+    /// Implementation of `spread` to use for this input
+    #[allow(unused)]
+    spread_l: Spread,
+    /// Implementation of `spread` to use for this input
+    #[allow(unused)]
+    spread_r: Spread,
```

Comment:
Thought: Could we have a `HeterogenousMerge` struct that manages the spreads? We could do something like:

```
let merger = HeterogeneousMerge::new(vec![schema1, schema2, schema3, ...]);
let merged = merger.merge(gathered_batches)?;
```

Reply:
(separately, not sure if the gatherer should be baked into that heterogeneous merge or not)